odoo/tools/lru.py

```diff
@@ -15,48 +14,99 @@ class LRU(MutableMapping[K, V], typing.Generic[K, V]):
+
+        # Initialize
         for key, value in pairs:
             self[key] = value
 
-    @locked
-    def __contains__(self, obj: K) -> bool:
-        return obj in self.d
+    @property
+    def count(self) -> int:
+        return self._count
+
+    def __contains__(self, key: object) -> bool:
+        return key in self._values
 
-    @locked
-    def __getitem__(self, obj: K) -> V:
-        a = self.d[obj]
-        self.d.move_to_end(obj, last=False)
-        return a
+    def __getitem__(self, key: K) -> V:
+        val = self._values[key]
+        # move key at the last position in self._ordering
+        self._ordering[key] = self._ordering.pop(key, None)
+        return val
 
-    @locked
-    def __setitem__(self, obj: K, val: V):
-        self.d[obj] = val
-        self.d.move_to_end(obj, last=False)
-        while len(self.d) > self.count:
-            self.d.popitem(last=True)
+    def __setitem__(self, key: K, val: V):
+        values = self._values
+        ordering = self._ordering
+        with self._lock:
+            values[key] = val
+            ordering[key] = ordering.pop(key, None)
+            while True:
+                # if we have too many keys in ordering, filter them out
+                if len(ordering) > len(values):
+                    # (copy to avoid concurrent changes on ordering)
+                    for k in ordering.copy():
+                        if k not in values:
+                            ordering.pop(k, None)
+                # check if we have too many keys
+                if len(values) <= self._count:
+                    break
+                # if so, pop the least recently used
+                try:
+                    # have a default in case of concurrent accesses
+                    key = next(iter(ordering), key)
+                except RuntimeError:
+                    # ordering modified during iteration, retry
+                    continue
+                values.pop(key, None)
+                ordering.pop(key, None)
 
-    @locked
-    def __delitem__(self, obj: K):
-        del self.d[obj]
+    def __delitem__(self, key: K):
+        self.pop(key)
 
-    @locked
     def __len__(self) -> int:
-        return len(self.d)
+        return len(self._values)
 
-    @locked
     def __iter__(self) -> Iterator[K]:
-        return iter(self.d)
+        with self._lock:
+            values = self._values
+            # remove keys for which there is no value
+            # (copy to avoid concurrent changes on ordering)
+            ordering = {k: None for k in self._ordering.copy() if k in values}
+            if len(values) > len(ordering):
+                # update with missing real key-values, we will just iterate once
+                ordering.update(values)  # type: ignore
+        return iter(ordering)
 
-    @locked
-    def pop(self, key: K) -> V:
-        return self.d.pop(key)
+    def pop(self, key: K, /, default=SENTINEL) -> V:
+        with self._lock:
```

Comment:
A small question, why not use `@locked`? for performance?

Reply:
The decorator `@locked` was introduced back then when there was no `with` statement in Python.  The `with` statement is just as clear, if not clearer IMHO.  Moreover, using `with self._lock` in a method and `@locked` for another method looks inconsistent to me.

Reply:
Do we prefer `with` than `@locked`? At least in this PR, we can use `@locked` for all sychronized methdos and remove one indent

Reply:
In my opinion `@locked` comes with some performance overhead and does not add any value here. It may have its uses elsewhere.